File: Fase 2/Evidencias Proyecto/proyecto/app ionic/TatasApp/src/app/interfaces/usuario.ts
import { Direccion } from "./direccion";

//interfaz para el registro de usuario
//creado por andrea el 19/04
export interface Usuario {
    mdl_nombres: string;
    mdl_apellidos: string;
    mdl_fecha_nacimiento: string; // Si prefieres, puedes usar `Date`, pero como lo ingresas desde un input tipo string, este es más flexible.
    mdl_correo_electronico: string;
    mdl_telefono: string;
    mdl_tipo_usuario: number;
    mdl_contrasena: string;
    mdl_confirmarContrasena: string;
    direccion: Direccion;
}

//interfaz para login
//creado por david el 20/04
export interface UsuarioLogin {
    correo: string;
    contrasena: string;
}

//interfaz para la respuesta de login exitoso
//creado por david el 21/04
export interface UsuarioLoginExitoso {
    id_usuario: number
    nombres: string
    tipo_usuario: number
    token: string
}

//interfaz para obtener los datos del usuario logueado
//creado por david el 28/04
export interface UsuarioLogueado {
    id_usuario: number
    nombres: string
    tipo_usuario: number
}

//interfaz para obtener los datos de usuario por id
//creado por david el 07/05
export interface UsuariosPorId {
    nombres: string
    apellidos: string
    fecha_nacimiento: Date
    correo: string
    tipo_usuario: number
    direccion_rel: Direccion
}

//interfaz para actualizar foto de perfil
//creado por david el 09/05
export interface FotoPerfil {
    id: number;
    foto_perfil: string;
}

//interfaz para editar datos de usuario
//creado por david el 09/05
export interface DatosUsuarioEditar {
    id: number;
    nombres: string;
    apellidos: string;
    fecha_nacimiento: string;
    telefono: string
    direccion: Direccion
}

//interfaz para editar correo
//creado por david el 10/05
export interface CorreoEditar {
    id: number;
    correo: string;
}

//interfaz para editar contraseña
//creado por david el 11/05
export interface ContrasenaEditar {
    id: number;
    contrasena: string;
}

File: Fase 2/Evidencias Proyecto/proyecto/api fastapi/app/schemas/usuario.py
#esquemas de usuario
#creado por david el 05/05

from pydantic import BaseModel
from typing import Optional
from datetime import date
from app.utils.helpers import (
    validador_no_string_vacio,
    validador_contrasena,
    validador_formato_correo,
    validador_formato_telefono
)


#REGISTRO DE USUARIO

#esquema para guardar direccion
#creado por david el 17/04
class DireccionCreate(BaseModel):
    direccion_texto: str
    adicional: Optional[str]

    #validador string vacio
    _validar_campos_str = validador_no_string_vacio(
        'direccion_texto'
    )

#esquema para guardar usuario completo
#creado por david el 17/04
class UsuarioCreate(BaseModel):
    nombres: str
    apellidos: str
    fecha_nacimiento: date
    correo: str
    telefono: str
    tipo_usuario: int
    contrasena: str
    direccion: DireccionCreate

    #validador string no vacio
    _validar_campos_str = validador_no_string_vacio('nombres', 'apellidos', 'correo', 'contrasena')

    #validador de formato contraseña
    _validar_contrasena = validador_contrasena(
        campo='contrasena',
        min_longitud=8,
        requerir_mayuscula=True,
        requerir_numero=True
    )

    #validador de formato correo
    _validar_formato_correo = validador_formato_correo('correo')

    #validador de formato telefono
    _validar_formato_telefono = validador_formato_telefono('telefono')

###########################################################################################################

#OBTENER INFO DE USUARIOS

#clases para mostrar el usuario con direccion
#creadas por david el 16/04
class DireccionOut(BaseModel):
    direccion_texto: str
    adicional: Optional[str]

    class Config:
        from_attributes = True

class UsuarioOut(BaseModel):
    nombres: str
    apellidos: str
    fecha_nacimiento: date
    correo: str
    telefono: str
    tipo_usuario: int #cambiado para mostrar el nombre del tipo_usuario
    direccion_rel: DireccionOut

    class Config:
        from_attributes = True

#esquema para hacer match con los contactos registrados en el telefono
#creado por david y andrea el 25/04
class ContactosRegistrados(BaseModel):
    id_usuario: int
    telefono: str
    tipo_usuario: int

#esquema para consultar la foto de perfil (separado de la consulta general de usuario)
#creado por david el 09/05
class FotoPerfilOut(BaseModel):
    id_usuario: int
    foto_perfil: Optional[str]

###########################################################################################

#LOGIN DE USUARIO

#esquema de datos para el login
#creado por david el 20/04
class UsuarioLogin(BaseModel):
    correo: str
    contrasena: str

#esquema para la respuesta del token
#creado por david el 20/04
class Token(BaseModel):
    access_token: str
    token_type: str = "bearer" #valor por defecto

#esquema para llenar el contenido de la respuesta de login exitoso
#creado por david el 20/04
class ContenidoLogin(BaseModel):
    id_usuario: int
    nombres: str
    tipo_usuario: int
    token: str

#esquema para la respuesta exitosa
#creado por david el 20/04
class RespuestaLoginExitoso(BaseModel):
    status: str = "success"
    message: str = "Credenciales válidas"
    contenido: ContenidoLogin

#esquema para la respuesta erronea
#creado por david el 20/04
class RespuestaLoginErronea(BaseModel):
    status: str = "error"
    message: str = "Credenciales inválidas"

#######################################################################################################

#ACTUALIZAR USUARIO

#esquema para editar la foto de perfil
#creado por david el 09/05
class FotoPerfilUpdate(BaseModel):
    id: int
    foto_perfil: Optional[str]

#esquemas para actualizar info de usuario (nombre y apellido, fecha nacimiento, telefono, direccion)
#creado por david el 09/05
class DireccionUpdate(BaseModel):
    direccion_texto: str
    adicional: Optional[str]

    _validar_campos_str = validador_no_string_vacio('direccion_texto')

class UsuarioUpdate(BaseModel):
    id: int
    nombres: str
    apellidos: str
    fecha_nacimiento: date
    telefono: str
    direccion: DireccionUpdate

    _validar_campos_str = validador_no_string_vacio('nombres', 'apellidos', 'telefono')
    _validar_formato_telefono = validador_formato_telefono('telefono')

#esquema para editar correo
#creado por david el 10/05
class CorreoUpdate(BaseModel):
    id: int
    correo: str

    _validar_campos_str = validador_no_string_vacio('correo')
    _validar_formato_correo = validador_formato_correo('correo')

#esquema para editar contraseña
#creado por david el 10/05
class ContrasenaUpdate(BaseModel):
    id: int
    contrasena: str

    _validar_campos_str = validador_no_string_vacio('contrasena')
    _validar_contrasena = validador_contrasena(
        campo='contrasena',
        min_longitud=8,
        requerir_mayuscula=True,
        requerir_numero=True
    )